packages/flutter_tools/test/runner/flutter_command_runner_test.dart

```diff
@@ -73,6 +80,36 @@ void main() {
       }, initializeFlutterRoot: false);
     });
 
+    group('version', () {
+      testUsingContext('checks that Flutter toJson output reports the flutter framework version', () async {
+        final ProcessResult result = ProcessResult(0, 0, 'random', '0');
+
+        when(processManager.runSync('git log -n 1 --pretty=format:%H'.split(' '),
+          workingDirectory: Cache.flutterRoot)).thenReturn(result);
+        when(processManager.runSync('git rev-parse --abbrev-ref --symbolic @{u}'.split(' '),
+          workingDirectory: Cache.flutterRoot)).thenReturn(result);
+        when(processManager.runSync('git rev-parse --abbrev-ref HEAD'.split(' '),
+          workingDirectory: Cache.flutterRoot)).thenReturn(result);
+        when(processManager.runSync('git ls-remote --get-url master'.split(' '),
+          workingDirectory: Cache.flutterRoot)).thenReturn(result);
+        when(processManager.runSync('git log -n 1 --pretty=format:%ar'.split(' '),
+          workingDirectory: Cache.flutterRoot)).thenReturn(result);
+        when(processManager.runSync('git describe --match v*.*.* --first-parent --long --tags'.split(' '),
+          workingDirectory: Cache.flutterRoot)).thenReturn(result);
```

Comment:
This will cause `GitTagVersion.determine()` to return `unknown`, which is the cause for the final version string to be "0.0.0-unknown".

The fact that it's going through this code path to get to the expected value will make this test very fragile.  I think this test can simply verify that we include the `'frameworkVersion'` key in the JSON, and that the value delegates to `frameworkVersion`.  To do this, you can write a local `FlutterVersion` implementation that extends from the real impl but overrides the `frameworkVersion` getter to return a known value.  Then in your test, you can simple expect that the requisite key exists in the JSON and that the matching value is your known value.

Reply:
Do you mean a FlutterVersion mock?

Reply:
Effectively, yes.  But unfortunately due to the way mocks work in mockito Dart, there's no way to say `when(...).thenCallRealMethod()`, which is what we really want.  If we had that, then we'd say:

```dart
when(mockVersion.frameworkVersion).thenReturn('foo');
when(mockVersion.toJson()).thenCallRealMethod();
Map<String, Object> json = mockVersion.toJson();
expect(json.containsKey('frameworkVersion'), isTrue);
expect(json['frameworkVersion'], 'foo');
```

Instead, we have to construct our own "mock" by extending `FlutterVersion` and overriding the `frameworkVersion` getter.

Reply:
ok yah that makes sense

Reply:
So it turns out you can't actually instance a `FlutterVersion` without giving those fake git responses. Because I don't think you can create a class and bypass the constructor?

Reply:
Yeah, you'll either need to (a) still fake the git responses, or (b) override all methods called by `toJson()`.

Reply:
Ok, so the fake git responses are fine? It's just the version response you didn't like?

Reply:
The fact that we're calling the real `frameworkVersion` getter makes the test fragile, which is what I don't like.  The best way forward is probably (b) ("override all methods called by `toJson()`").  Something like:

```dart
class FlutterVersionForToJson extends FlutterVersion {
  @override
  String get frameworkVersion = 'frameworkVersion';

  @override
  String get channel => 'channel';

  @override
  String get repositoryUrl => 'repositoryUrl';

  @override
  String get frameworkRevision => 'frameworkRevision';

  @override
  String get frameworkCommitDate => 'frameworkCommitDate';

  @override
  String get engineRevision => 'engineRevision';

  @override
  String get dartSdkVersion => 'dartSdkVersion';
}
```

Then you won't need the mock git stuff at all.

Reply:
But I think you will because I don't think you can initialize anything that inherits from `FlutterVersion` without going through that git stuff. Just to clarify, right now the mock git stuff doesn't have any meaningful response, it merely exists to not cause references to be null.